Fix tests warning and deprecations

ZRazor opened the pull request:
I am targeting this branch, because it is BC.

## Changelog

```markdown

### Fixed
- Fixed tests namespaces
- Replaced deprecated `getMock` with `createMock`
- Added support for FOSRestBundle >= 2.0
- Replaced deprecated `BaseBlockService`
- Replaced deprecated `$form->bind($request)` with `$form->handleRequest($request)`
- Replaced twig deprecated `renderMetadatas ` with `getMetadatas `
- Increased block-bundle dependency to ^3.2
```

## Subject

#### Before
```
.WWWWWWWWWWWWWWWWWWWW..W.WWWWWWWWWWWWWWWWWWWWWWWSSSWWWWWWW.WWWW  63 / 327 ( 19%)
WWWWWWW..........WW..WWWWW.....W.WWWWWWWW...W.WWWW..WW.WWWWWWW. 126 / 327 ( 38%)
WWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWWW...WWWWWWWWW.WW..WWWWWWW 189 / 327 ( 57%)
WW.WWWWWWWWWWWWWWWWWWWWWWWWSSS.SSSSS.WWWW.WWWW.SSSSSS.WWWW.WWWW 252 / 327 ( 77%)
SSSSSSSSS.WW.......WWWWWWWWWWWWWWWW.WWSSSSSWWW.WWWWWWWWWWWWWWWW 315 / 327 ( 96%)
WWWWWWWWW.W.                                                    327 / 327 (100%)
WARNINGS!
Tests: 327, Assertions: 924, Warnings: 240, Skipped: 31.

Remaining deprecation notices (55)
```
#### After
```
...............................................................  63 / 327 ( 19%)
............................................................... 126 / 327 ( 38%)
............................................................... 189 / 327 ( 57%)
............................................................... 252 / 327 ( 77%)
............................................................... 315 / 327 ( 96%)
............                                                    327 / 327 (100%)
OK (327 tests, 1002 assertions)
```

jordisala1991 commented on src/BasketBundle/Block/BasketBlockService.php:
```diff
@@ -22,7 +22,7 @@
  * @author Xavier Coureau <xcoureau@ekino.com>
  */
-class BasketBlockService extends BaseBlockService
+class BasketBlockService extends AbstractBlockService
```
Care with this one. It should be abstractadminblockservice

ZRazor replied:
Thanks, I'll change it

OskarStark replied:
are you sure @jordisala1991 i think this is the basket block for the frontend users

jordisala1991 replied:
I am not sure if it was an error but take a look: https://github.com/sonata-project/SonataBlockBundle/blob/3.x/Block/BaseBlockService.php

jordisala1991 replied:
@core23 can you help with this one?

jordisala1991 replied:
@greg0ire do you know the answer? It should be AbstractBlockService or AbstractAdminBlockService?

greg0ire replied:
It's clearly an admin block, so the latter.

jordisala1991 commented on src/BasketBundle/Controller/Api/BasketController.php:
```diff
@@ -355,10 +355,9 @@ public function deleteBasketBasketelementsAction($basketId, $elementId)
-        $serializationContext->setGroups(array('sonata_api_read'));
-        $serializationContext->enableMaxDepthChecks();
-        $view->setSerializationContext($serializationContext);
+        $context = new Context();
```
Is this BC with the older versions of FosRestBundle?

ZRazor replied:
At the current moment in 2.x there are `"friendsofsymfony/rest-bundle": "^1.1"` and `"sonata-project/media-bundle": "^3.0"`, but SonataMediaBundle 3.0.0 requires `"friendsofsymfony/rest-bundle": "^1.7.4 || ^2.0",` so there is no support for older rest-bundle versions.
Thats why I increased to `"friendsofsymfony/rest-bundle": "^1.8"` and removed `SerializationContext`, because it is removed in rest-bundle 2.0

ZRazor replied:
But it definitely wont work with FOSRestBundle 1.1. If this is a BC, i will target master branch.

jordisala1991 replied:
Take a look at this: https://github.com/sonata-project/SonataMediaBundle/blob/3.x/Controller/Api/MediaController.php#L410

OskarStark commented on src/CustomerBundle/Block/RecentCustomersBlockService.php:
```diff
@@ -25,7 +25,7 @@
  * @author Hugo Briand <briand@ekino.com>
  */
-class RecentCustomersBlockService extends BaseBlockService
+class RecentCustomersBlockService extends AbstractBlockService
```
this needs to extend from: `AbstractAdminBlockService`, as it is for the backend

OskarStark commented on tests/Component/Basket/BasketBuilderTest.php:
```diff
@@ -57,25 +58,25 @@ public function testBuildWithNonExistentProductCode()
         $basketBuilder = new BasketBuilder($productPool, $addressManager, $deliveryPool, $paymentPool);
 
-        $basketElement_1 = $this->getMock('Sonata\Component\Basket\BasketElementInterface');
+        $basketElement_1 = $this->createMock('Sonata\Component\Basket\BasketElementInterface');
```
can you fix this variable name to `$BasketElement1` as we don't use underscores for variable names, thank you

OskarStark commented on tests/Component/Basket/BasketBuilderTest.php:
```diff
@@ -85,18 +86,18 @@ public function testBuild()
         $basketBuilder = new BasketBuilder($productPool, $addressManager, $deliveryPool, $paymentPool);
 
-        $basketElement_1 = $this->getMock('Sonata\Component\Basket\BasketElementInterface');
+        $basketElement_1 = $this->createMock('Sonata\Component\Basket\BasketElementInterface');
```
same here

OskarStark commented on tests/Component/Currency/CurrencyDetectorTest.php:
```diff
@@ -9,18 +9,19 @@
+use Sonata\Tests\Helpers\PHPUnit_Framework_TestCase;
 
 /**
  * Test class for CurrencyDetector.
```
please remove this comment, its useless

OskarStark commented on tests/Component/Currency/CurrencyManagerTest.php:
```diff
@@ -9,20 +9,21 @@
+use Sonata\Tests\Helpers\PHPUnit_Framework_TestCase;
 
 /**
  * Test class for CurrencyManager.
```
please remove this comment, its useless

OskarStark commented on tests/Component/Payment/CheckPaymentTest.php:
```diff
@@ -27,22 +28,22 @@ public function getId()
+class CheckPaymentTest extends PHPUnit_Framework_TestCase
 {
     /**
      * useless test ....
```
lol? 😄

OskarStark commented on tests/Component/Payment/Scellius/ScelliusPaymentTest.php:
```diff
@@ -100,21 +101,21 @@ public function testValidPayment()
         $transaction->expects($this->any())->method('get')->will($this->returnValue('" >> /dev/null'));
         $transaction->expects($this->any())->method('getParameters')->will($this->returnValue(array()));
-        $transaction->expects($this->any())->method('getOrder')->will($this->returnValue($this->getMock('Sonata\Component\Order\OrderInterface')));
+        $transaction->expects($this->any())->method('getOrder')->will($this->returnValue($this->createMock('Sonata\Component\Order\OrderInterface')));
```
this line is very long, please linebreak it

OskarStark commented on tests/Component/Product/BaseProductServiceTest.php:
```diff
@@ -124,20 +125,20 @@ class BaseOrderElementTest_ProductProvider extends BaseOrderElement
-        $basketElementManager = $this->getMock('\Sonata\Component\Basket\BasketElementManagerInterface');
-        $basketElementManager->expects($this->any())->method('getClass')->will($this->returnValue('\Sonata\Tests\Component\Product\BaseOrderElementTest_ProductProvider'));
+        $basketElementManager = $this->createMock('\Sonata\Component\Basket\BasketElementManagerInterface');
+        $basketElementManager->expects($this->any())->method('getClass')->will($this->returnValue('\Sonata\Component\Tests\Product\BaseOrderElementTest_ProductProvider'));
```
please linebreak this

OskarStark commented on tests/CustomerBundle/Controller/Api/CustomerControllerTest.php:
```diff
@@ -48,34 +49,34 @@ public function testGetCustomerActionNotFoundException()
+        $customer = $this->createMock('Sonata\Component\Customer\CustomerInterface');
+        $order = $this->createMock('Sonata\Component\Order\OrderInterface');
 
         $this->assertEquals(array($order), $this->createCustomerController($customer, null, null, null, $order)->getCustomerOrdersAction(1));
```
please linebreak this

OskarStark commented on tests/CustomerBundle/Controller/Api/CustomerControllerTest.php:
```diff
@@ -104,21 +105,21 @@ public function testPostCustomerInvalidAction()
+        $formFactory = $this->createMock('Symfony\Component\Form\FormFactoryInterface');
         $formFactory->expects($this->once())->method('createNamed')->will($this->returnValue($form));
 
         $view = $this->createCustomerController($customer, $customerManager, $addressManager, $formFactory)->postCustomerAddressAction(1, new Request());
```
please linebreak this

OskarStark commented on tests/CustomerBundle/Controller/Api/CustomerControllerTest.php:
```diff
@@ -147,18 +148,18 @@ public function testPostCustomerAddressInvalidAction()
+        $formFactory = $this->createMock('Symfony\Component\Form\FormFactoryInterface');
         $formFactory->expects($this->once())->method('createNamed')->will($this->returnValue($form));
 
         $view = $this->createCustomerController($customer, $customerManager, null, $formFactory)->putCustomerAction(1, new Request());
```
please linebreak this

OskarStark commented on tests/CustomerBundle/Controller/Api/CustomerControllerTest.php:
```diff
@@ -168,17 +169,17 @@ public function testPutCustomerAction()
+        $formFactory = $this->createMock('Symfony\Component\Form\FormFactoryInterface');
         $formFactory->expects($this->once())->method('createNamed')->will($this->returnValue($form));
 
         $view = $this->createCustomerController($customer, $customerManager, null, $formFactory)->putCustomerAction(1, new Request());
```
please linebreak this

OskarStark commented on tests/Helpers/PHPUnit_Framework_TestCase.php:
```diff
@@ -0,0 +1,54 @@
+/**
+ * This is helpers class for supporting old and new PHPUnit versions.
+ *
+ * @todo NEXT_MAJOR: Remove this class when dropping support for < PHPUnit 5.4.
```
IMO `@todo` is not needed here. @greg0ire @jordisala1991 @core23

jordisala1991 replied:
Agree, this is not needed, but I don't know if this was copied from other bundle. If that is the case, this is wrong on a lot of bundles. Let me check:

https://github.com/sonata-project/SonataMediaBundle/blob/3.x/Tests/Helpers/PHPUnit_Framework_TestCase.php

We could remove it here, but it should not matter because we will drop this code soon on next major

ZRazor replied:
Yep, I copied it from AdminBundle, where it was originally comitted.
I rebased and removed it.

OskarStark commented on tests/ProductBundle/Controller/Api/ProductControllerTest.php:
```diff
@@ -48,86 +49,86 @@ public function testGetProductActionNotFoundException()
+        $productCategory = $this->createMock('Sonata\Component\Product\ProductCategoryInterface');
         $product->expects($this->once())->method('getProductCategories')->will($this->returnValue(array($productCategory)));
 
         $this->assertEquals(array($productCategory), $this->createProductController($product)->getProductProductcategoriesAction(1));
```
please linebreak this

OskarStark commented on tests/ProductBundle/Controller/Api/ProductControllerTest.php:
```diff
@@ -48,86 +49,86 @@ public function testGetProductActionNotFoundException()
+        $productCollection = $this->createMock('Sonata\Component\Product\ProductCollectionInterface');
         $product->expects($this->once())->method('getProductCollections')->will($this->returnValue(array($productCollection)));
 
         $this->assertEquals(array($productCollection), $this->createProductController($product)->getProductProductcollectionsAction(1));
```
please linebreak this

OskarStark commented on tests/ProductBundle/Controller/Api/ProductControllerTest.php:
```diff
@@ -48,86 +49,86 @@ public function testGetProductActionNotFoundException()
+        $formFactory = $this->createMock('Symfony\Component\Form\FormFactoryInterface');
         $formFactory->expects($this->once())->method('createNamed')->will($this->returnValue($form));
 
         $view = $this->createProductController(null, $productManager, $productPool, $formFactory, $formatterPool)->postProductAction('my.test.provider', new Request());
```
please linebreak this

OskarStark commented on tests/ProductBundle/Controller/Api/ProductControllerTest.php:
```diff
@@ -137,22 +138,22 @@ public function testPostProductAction()
+        $formFactory = $this->createMock('Symfony\Component\Form\FormFactoryInterface');
         $formFactory->expects($this->once())->method('createNamed')->will($this->returnValue($form));
 
         $view = $this->createProductController(null, $productManager, $productPool, $formFactory, $formatterPool)->postProductAction('my.test.provider', new Request());
```
please linebreak this

OskarStark commented on tests/ProductBundle/Controller/Api/ProductControllerTest.php:
```diff
@@ -162,24 +163,24 @@ public function testPostProductInvalidAction()
+        $formFactory = $this->createMock('Symfony\Component\Form\FormFactoryInterface');
         $formFactory->expects($this->once())->method('createNamed')->will($this->returnValue($form));
 
         $view = $this->createProductController($product, $productManager, $productPool, $formFactory, $formatterPool)->putProductAction(1, 'my.test.provider', new Request());
```
please linebreak this

OskarStark commented on tests/ProductBundle/Controller/Api/ProductControllerTest.php:
```diff
@@ -189,23 +190,23 @@ public function testPutProductAction()
+        $formFactory = $this->createMock('Symfony\Component\Form\FormFactoryInterface');
         $formFactory->expects($this->once())->method('createNamed')->will($this->returnValue($form));
 
         $view = $this->createProductController($product, $productManager, $productPool, $formFactory, $formatterPool)->putProductAction(1, 'my.test.provider', new Request());
```
please linebreak this

OskarStark commented on tests/ProductBundle/Seo/Services/FacebookTest.php:
```diff
@@ -43,22 +44,16 @@ public function testAlterPage()
         $numberHelper = $this->getMockBuilder('Sonata\IntlBundle\Templating\Helper\NumberHelper')->disableOriginalConstructor()->getMock();
         $currencyDetector = $this->getMockBuilder('Sonata\Component\Currency\CurrencyDetectorInterface')->disableOriginalConstructor()->getMock();
```
we can use `createMock` here too

OskarStark commented on tests/ProductBundle/Seo/Services/FacebookTest.php:
```diff
@@ -43,22 +44,16 @@ public function testAlterPage()
         $numberHelper = $this->getMockBuilder('Sonata\IntlBundle\Templating\Helper\NumberHelper')->disableOriginalConstructor()->getMock();
```
we can use `createMock` here too

OskarStark commented on tests/Component/Form/BasketValidatorTest.php:
```diff
@@ -9,22 +9,23 @@
+        $provider = $this->createMock('Sonata\Component\Product\ProductProviderInterface');
         $provider->expects($this->once())->method('validateFormBasketElement');
 
         $pool = $this->getMockBuilder('Sonata\Component\Product\Pool')->disableOriginalConstructor()->getMock();
```
we can use `createMock` here too

OskarStark commented on tests/Component/Form/EventListener/BasketResizeFormListenerTest.php:
```diff
@@ -33,12 +34,12 @@ public function testPreSetData()
     {
         $builder = $this->getMockBuilder('Symfony\Component\Form\FormBuilder')->disableOriginalConstructor()->getMock();
```
we can use `createMock` here too

ZRazor commented:
@OskarStark @jordisala1991 

- Remove commit with composer.json update - done
- Add compatibility with old versions of FOSRestBundle - done
- Replace all `this->getMockBuilder\((.*)\)->disableOriginalConstructor\(\)->getMock\(\)` with `this->createMock($1)` - done
- Other notes - done
- Fix changelog - done

Also found an issue in `PHPUnit_Framework_TestCase` wrapper. Calling `getMock` instead of `createMock` is not equal. And when I replaced `getMockBuilder('')-> disableOriginalConstructor()->getMock()` I got tonns of errors in travis on php 5.3 This is beacause `getMock` in PHPUnit 4 does not disable original constructor. So I fixed the wrapper by replacing `parent::getMock` with close implelentation of `createMock` from https://github.com/sebastianbergmann/phpunit/blob/c465bb99341f11bd0f6bdba0e2338b50fa7a7218/src/Framework/TestCase.php#L1441

greg0ire commented:
Great job with the "polyfill" of createMock @ZRazor !

greg0ire commented on composer.json:
```diff
@@ -65,15 +65,16 @@
+            "Sonata\\PaymentBundle\\Tests\\": "tests/PaymentBundle",
+            "Sonata\\PriceBundle\\Tests\\": "tests/PriceBundle",
+            "Sonata\\ProductBundle\\Tests\\": "tests/ProductBundle",
+            "Sonata\\Tests\\Helpers\\": "tests/Helpers"
```
Not sure why anyone would need to autoload tests, but 👍

greg0ire replied:
And if they do, the namespace change is a BC-break for them. IMO, you could just remove the autoload-dev section.

greg0ire commented on phpunit.xml.dist:
```diff
@@ -9,7 +9,7 @@
          stopOnFailure="false"
          syntaxCheck="false"
-         bootstrap="tests/bootstrap.php"
+         bootstrap="tests/tests/bootstrap.php"
```
This doesn't look too good… can you find a better name for the nested directory?

ZRazor replied:
@greg0ire, actually it is the way how it is done in other sonata bundles

greg0ire replied:
Ok leave it like that then… but… still quite weird if you ask me 😅

ZRazor replied:
I agree, but I checked admin, media, classification - all have the same structure `bootstrap="Tests/tests/bootstrap.php"`   ¯\\_(ツ)_/¯
Thats is also true for above note, all these bundles have autoload-dev with tests in composer.json
I suppose that's because namespaces in tests are not PSR-4 compliant

greg0ire replied:
Only the fixtures might need to be autoloadable… tests that should be used outside the bundle should not be in tests, they should be in `src/Test`… but anyway, I thought about it more, it's not a BC break, b/c this is autoload-dev

greg0ire commented:
@OskarStark @jordisala1991 please review again

jordisala1991 commented:
If we are supporting FOSRestBundle 2.x, can you add it to composer.json?

ZRazor commented:
@jordisala1991, Done

jordisala1991 commented:
Let's merge @greg0ire ?

@OskarStark issues are already solved

greg0ire commented:
Thanks a lot @ZRazor !

jordisala1991 commented:
Amazing work @ZRazor
